[POST] Visual thinking (#57)
allegro/blog: _posts/2022-02-03-visual-thinking.md

Do we use the most optimal way to express our ideas?

## The Story

### Spoken planning

— Josh, could you summarize what we are going to implement, please?

That was a bit unexpected. He was just trying to match what he already knew from onboarding days with what was just
said. It was really difficult to follow team discussion at the same time. It was a standard planning session, held over
Zoom, with his distributed team. Someone was already sharing screen and the story summary, along with acceptance
criteria, was visible. They were developing an online store and the current topic was: basket price reduction for active
users. In short: users who spent more than 50€ for the last 7 days should have a discount applied during the checkout
process.

— Well, we are going to add a new service which is going to hold this discount logic. We will provide an API for the
cart service to call us and we will check the transactions store for recent orders.<br/>
— Is that all?<br/>
— I didn’t catch more changes.<br/>
— What about customers willing to check if they are eligible for a discount?<br/>

> — Oh, so it seems we need to change ‘my account’ page as well.<br/>

Reviewer: ```suggestion
— Oh, so it seems we need to change the “My Account” page as well.<br/>
```

— And the checkout service? We have to both display discount and use it.<br/>
— You’re right! I was trying to picture the main change in my mind and wasn’t paying attention to the whole
discussion.<br/>
— Guys, checkout stays the same. Cart is providing everything the checkout requires.

### Retrospective

Let’s stop here for a moment. Have you ever been in a situation, when you were forced to do two things at the same time?

> For example listening to what is being said and trying to actively participate in the discussion regarding a not exactly

Reviewer 2: ```suggestion
For example, listening to what is being said and trying to actively participate in the discussion regarding a not exactly
```

> well known topic? Was it a demanding experience? I had such an opportunity and I remember these sessions as quite

Reviewer 2: ```suggestion
well-known topic? Was it a demanding experience? I had such an opportunity and I remember these sessions as quite
```

> demanding. Usually after a hourly session I was exhausted and in need of a break. Not to mention that it required a

Reviewer 2: ```suggestion
demanding. Usually, after an hourly session, I was exhausted and in need of a break. Not to mention that it required a
```
To jest ciekawe, bo mimo że mamy spółgłoskę to z powodu brzmienia lepiej pasuje tutaj `an`.

significant amount of writing to capture everything that was said — just to have an option of referring back to this
during the sprint. What if this session looked differently?

### Visual planning

> “Josh, could you summarize what we are going to implement?”.

Reviewer 2: Tutaj specjalnie inna forma niż za pierwszym razem w linijce 16?

Author: Raczej niedopatrzenie. Zrobię identycznie, to powinno dać ładną kompozycję przez powtórzenie

> That was a bit unexpected, however it was a no-brainer.

Reviewer 2: ```suggestion
That was a bit unexpected, however, it was a no-brainer.
```

![planning services](/img/articles/2022-02-03-visual-thinking/planning_services.png)
“As we can see in the picture we are going to add a new service, discounts. This service will be called by the current

> cart service to display reduced price (if applicable). Also the “My Account” page is going to call us to retrieve the

Reviewer 2: ```suggestion
cart service to display a reduced price (if applicable). Also, the “My Account” page is going to call us to retrieve the
```

current rebate for the logged user”.

### The difference

> What is the main difference between the first scenario and the second one? To me it is about common model. In the first

Reviewer 2: ```suggestion
What is the main difference between the first scenario and the second one? To me, it is about a common model. In the first
```

> case I am building an individual model of a change. I need significant amount of energy to translate speech to my model

Reviewer 2: ```suggestion
case, I am building an individual model of a change. I need a significant amount of energy to translate speech to my model
```

> and to explain my model to the others. What is more — everyone involved in the discussion is building a mental model on

Reviewer 2: ```suggestion
and to explain my model to others. What is more — everyone involved in the discussion is building a mental model on
```

their own with a similar amount of energy spent on synchronization.

![private models](/img/articles/2022-02-03-visual-thinking/private_models.png)

> In case of the second scenario the model is common. There is no need to maintain private models. It is easy to

Reviewer 2: ```suggestion
In the case of the second scenario, the model is common. There is no need to maintain private models. It is easy to
```
Zastanawiam się też czy `shared` by tutaj lepiej nie pasowało niż `common`.

Author: shared brzmi bardzo na miejscu

understand changed elements and to refer to discussed changes later. “A picture is worth a thousand words” after all.

![common model](/img/articles/2022-02-03-visual-thinking/common_model.png)

Output from a visual planning is something that can be used further during a sprint. Depending on the tool it is
possible to use it instead of an issue-tracking tool. Sticky notes can indicate actions, tasks, TODOs. They can be

> arranged in a tree and display scope of a pull-request. They can be connected by a dotted line to indicate dependencies.

Reviewer 2: ```suggestion
arranged in a tree and display the scope of a pull request. They can be connected by a dotted line to indicate dependencies.
```

And in the worst case, when you do not have any idea how to visualize something, you can always use a block of text and
describe it using words.

![planning tasks](/img/articles/2022-02-03-visual-thinking/planning_tasks.png)

## The background

Why switching to a drawing board has such an effect? To answer this question we have to check some facts.

> On a daily basis a spoken (or written) language is our standard way of communication. Children, however, need some time

Reviewer 2: ```suggestion
On a daily basis, a spoken (or written) language is our standard way of communication. Children, however, need some time
```

> to develop such skill. Earlier they are able to:

Reviewer: ```suggestion
to develop such a skill. Earlier they are able to:
```

* register movement (at 2 months age)
* try to grab things by hand (5 months)

> * follow movement with our eyes, find hidden things (7 months)

Reviewer: ```suggestion
* follow movement with their eyes, find hidden things (7 months)
```
Zaczynamy akapit od "Earlier they are able to:"

> * exploit cause-effect — drop a toy an watch it falling (8 months)

Reviewer: ```suggestion
* exploit cause-effect — drop a toy and watch it fall (8 months)
```

* start to use words in a proper context (11 months)

> Such spatial skills are something all creatures need to develop to survive. Even plants, to some extent, exibit

Reviewer 2: ```suggestion
Such spatial skills are something all creatures need to develop to survive. Even plants, to some extent, exhibit
```

spatial-aware behavior — they move to follow the sun.

It is worth noting that spatial-related terms
like [“where, here, near, etc.”](https://en.wikipedia.org/wiki/Natural_semantic_metalanguage)
can be translated to any language in the world.

On the other hand, what language do we use to express ideas-related actions? As Barbara Tversky listed in her
[“Mind in motion”](https://www.youtube.com/watch?v=gmc4wEL2aPQ) lecture we can:

* raise ideas
* pull them together
* tear apart
* turn inside out
* push forward
* toss out

![ideas and actions](/img/articles/2022-02-03-visual-thinking/ideas_and_actions.png)

We talk about ideas in the same way as about any space-related topic!

> How could it happen? Scientists have been trying to understand functions of different parts of the brain for some time

Reviewer 2: ```suggestion
How could it happen? Scientists have been trying to understand the functions of different parts of the brain for some time
```

> already. In the seventies they have identified so-called **place cells** — neurons that are activated at specific

Reviewer 2: ```suggestion
already. In the seventies, they have identified so-called **place cells** — neurons that are activated at a specific
```

> already. In the seventies, they have identified so-called **place cells** — neurons that are activated at a specific

Reviewer: ```suggestion
already. In the seventies, they have identified the so-called **place cells** — neurons that are activated at a specific
```

location. It took some time to identify another layer of neurons on top of these: **grid cells**, working as our inner
GPS, activated when switching locations. The latter discovery was

> awarded [Nobel Prize in 2014](https://www.nobelprize.org/prizes/medicine/2014/press-release/).

Reviewer: ```suggestion
awarded a [Nobel Prize in 2014](https://www.nobelprize.org/prizes/medicine/2014/press-release/).
```

Tests on human beings resulted in another finding: place cells are activated not only in specific locations. They are

> also activated by events, people, ideas. The grid cells are activated by thinking about consequences of events, by

Reviewer 2: ```suggestion
also activated by events, people, ideas. The grid cells are activated by thinking about the consequences of events, by
```

> social interactions and by connecting ideas together. We are using the same brain structures for spatial orientation,

Reviewer 2: ```suggestion
social interactions, and by connecting ideas together. We are using the same brain structures for spatial orientation,
```

> for ideas and for social interactions. This is the reason why Barbara Tversky issued an audacious thesis that “all

Reviewer 2: ```suggestion
for ideas, and for social interactions. This is the reason why Barbara Tversky issued an audacious thesis that “all
```

> thoughts begins as spatial thoughts”.

Reviewer 2: ```suggestion
thoughts begin as spatial thoughts”.
```

### Map elements

Nowadays we have a GPS sensor in almost any smartphone, so it is rather difficult to get lost. Basic GPS information -
current coordinate — is not very useful by itself. It is much more convenient to display our current location over a map

> layer. What is a map? According to [wikipedia](https://en.wikipedia.org/wiki/Map) it is “[..] a symbolic depiction

Reviewer 2: ```suggestion
layer. What is a map? According to [wikipedia](https://en.wikipedia.org/wiki/Map), it is “[..] a symbolic depiction
```

emphasizing relationships between elements of some space [..]”.

As our ideas — and imaginary concepts like services — are treated by our brains as spatial elements we simply use known
concepts of space visualization to present imagined beings and relations among them.

### Arrows

> We used so far several symbols in our planning diagrams. Almost all of them can be found in maps as well, except one.

Reviewer: ```suggestion
So far we used several symbols in our planning diagrams. Almost all of them can be found in maps as well, except one.
```
Tak brzmi mi to naturalniej.

This element is an arrow — a significant element of our drawings. According to Barbara Tversky in the already
mentioned [“Mind in motion” lecture](https://www.youtube.com/watch?v=gmc4wEL2aPQ)
arrows as visual elements started to appear in the 20th century. Before that symbols of feet or fingers had been used to

> indicate direction. Addition of arrows changes our perception of diagrams: without arrows it is usually a structural

Reviewer 2: ```suggestion
indicate direction. The addition of arrows changes our perception of diagrams: without them, it is usually a structural
```

> drawing, that requires additional labels to understand it. Arrows transform such a structural diagram into functional

Reviewer 2: ```suggestion
drawing, that requires additional labels to understand. Arrows transform such a structural diagram into a functional
```

> diagram: we are able to trace arrows to their origin, we see how things are connected and how they cooperate. Check

Reviewer 2: ```suggestion
diagram: we can trace arrows to their origin, we see how things are connected and how they cooperate. Check the
```

> example below — left-hand side version is static, only describes elements and the right-hand side version shows movement

Reviewer 2: ```suggestion
example below — the left-hand side version is static, only describes elements and the right-hand side version shows the movement
```

of particular elements without using a single word.

![static and dynamic clock](/img/articles/2022-02-03-visual-thinking/clock.png)

## Messy lines

> All these well-known elements are our means of communication with other people, or even ourselves, in a different point

Reviewer 2: ```suggestion
All these well-known elements are our means of communication with other people, or even ourselves, at a different point
```

> in time. Sometimes we draw to discover, we sketch shapes to find an inspiration, an idea. This seems to be important in

Reviewer 2: ```suggestion
in time. Sometimes we draw to discover, we sketch shapes to find inspiration, an idea. This seems to be important in
```

> a different creative profession: architecture. Architects discover ideas in sketches. Ambiguity of non-obvious shapes

Reviewer 2: ```suggestion
a different creative profession: architecture. Architects discover ideas in sketches. The ambiguity of non-obvious shapes
```
